### PR TITLE
- generate challenge randomness

### LINKED ISSUE #1167: Generate Challenge - randomness
- While populating BlobberChallenge partition, we should also store - data it's actively storing.
- To increase randomness, select n blobberChallenge partition and 1 blobber from each of partition.
- from selected n blobber, weight based on data stored in Bi
- select the blobber with maximum weight and continue the existing algorithm from there

The intent of this change should be:
- blobbers storing more data should have higher chance of getting rewarded
- increase randomness in the static partitions

This will be extension to #1055

### PR DESCRIPTION
**Pull request by ssardana08**
- generate challenge weight

## Tests 
Tasks to complete before merging PR:
- [ ]  Ensure system tests are passing. If not [Run them manually](https://github.com/0chain/0chain/actions/workflows/system_tests.yml) to check for any regressions :clipboard:
- [ ]  Do any new system tests need added to test this change? do any existing system tests need updated? If so create a PR at [0chain/system_test](https://github.com/0chain/system_test)
- [ ]  Merge your system tests PR to master AFTER merging this PR

## Associated PRs (Link as appropriate):
- blobber:
- gosdk:
- system_test:
- zboxcli:
- zwalletcli:
- Other: ...

### CODE REVIEW

File: code/go/0chain.net/smartcontract/storagesc/challenge.go [hunk 1]
```diff
@@ -704,17 +704,27 @@ func (sc *StorageSmartContract) populateGenerateChallenge(
 	balances c_state.StateContextI,
 ) (*challengeOutput, error) {
 
+	const maxBCPartitionSelect = 5
```
Comment (peterlimg) [inline code/go/0chain.net/smartcontract/storagesc/challenge.go]: Rename to maxBlobbersSelect perhaps as we don't select partitions anymore.

File: code/go/0chain.net/smartcontract/storagesc/challenge.go [hunk 2]
```diff
@@ -704,17 +704,27 @@ func (sc *StorageSmartContract) populateGenerateChallenge(
...
 			"error getting random slice from blobber challenge partition")
 	}
 
-	randomIndex := challRand.Intn(len(blobberChallenges))
-	bcItem := blobberChallenges[randomIndex]
-	Logger.Debug("generate_challenges", zap.Int("random index", randomIndex),
-		zap.String("blobber id", bcItem.BlobberID), zap.Int("blobber challenges", len(blobberChallenges)))
+	for i := 0; i < maxBCPartitionSelect; i++ {
+		randomIndex := challRand.Intn(len(blobberChallenges))
+		tempItem := blobberChallenges[randomIndex]
```
Comment (peterlimg) [inline code/go/0chain.net/smartcontract/storagesc/challenge.go]: We don't have to get random item if the total blobber number is less than 5, in that case we can iterate through the blobberChallenges directly.

**Review by peterlimg — CHANGES_REQUESTED** [review]

File: code/go/0chain.net/smartcontract/storagesc/challenge.go (hunk 2 above)
Comment (ssardana08) [reply code/go/0chain.net/smartcontract/storagesc/challenge.go]: still in partition size of 50, we need to select 5 blobbers randomly, no?

Comment (peterlimg) [reply code/go/0chain.net/smartcontract/storagesc/challenge.go]: Think about the case that partition size of 50, total blobbers is 4, maxBCPartitionSelect is 5, do we need to get the 4 items 5 times randomly? No, we can iterate through the blobbers list directly. This is a corner case which may not happen in really, but we need to know it.

Comment (ssardana08) [reply code/go/0chain.net/smartcontract/storagesc/challenge.go]: So if len(partition) <= maxBCpartitionSelect, we should iterate.. else random select, right?

Comment (peterlimg) [reply code/go/0chain.net/smartcontract/storagesc/challenge.go]: yes

File: code/go/0chain.net/smartcontract/storagesc/challenge.go [hunk 3]
```diff
@@ -716,13 +716,20 @@ func (sc *StorageSmartContract) populateGenerateChallenge(
...
-		tempItem := blobberChallenges[randomIndex]
+	var blobbersSelected []BlobberChallengeNode
+	if len(blobberChallenges) <= maxBlobbersSelect {
+		blobbersSelected = blobberChallenges
+	} else {
+		for i := 0; i < maxBlobbersSelect; i++ {
+			randomIndex := challRand.Intn(len(blobberChallenges))
+			blobbersSelected = append(blobbersSelected, blobberChallenges[randomIndex])
+		}
+	}
```
Comment (peterlimg) [inline code/go/0chain.net/smartcontract/storagesc/challenge.go:727]: Init the `blobbersSelected` with capacity.

**Review by peterlimg — CHANGES_REQUESTED** [review]

**Review by peterlimg — APPROVED** [review]
lgtm

**Comment by service-0chain** [comment]
Manual system tests [success] with the following config
| config  | value |
| ------------- | ------------- |
| tests skipped  | FALSE  |
| system-tests  | master  |
| gosdk  | NOT OVERRIDDEN  |
| zboxcli  | staging  |
| zwalletcli  | staging  |
| 0chain  | staging  |
| blobber  | staging  |
| authorizer  | staging  |
| 0box  | staging  |
| 0block  | staging  |
| 0dns  | staging  |
| explorer  | master  |
| 0proxy  | staging  |
| 0search  | staging  |

**Pull request merged**